File: packages/drape-vision/android/src/main/java/com/margelo/nitro/drape/vision/HybridDrapeSpecialistLandmarkers.kt
package com.margelo.nitro.drape.vision

import android.content.Context
import android.graphics.Bitmap
import com.google.mediapipe.framework.image.BitmapExtractor
import com.google.mediapipe.framework.image.BitmapImageBuilder
import com.google.mediapipe.framework.image.ByteBufferExtractor
import com.google.mediapipe.framework.image.MPImage
import com.google.mediapipe.tasks.components.containers.Category
import com.google.mediapipe.tasks.components.containers.Landmark
import com.google.mediapipe.tasks.components.containers.NormalizedLandmark
import com.google.mediapipe.tasks.core.BaseOptions
import com.google.mediapipe.tasks.core.Delegate
import com.google.mediapipe.tasks.vision.core.RunningMode
import com.google.mediapipe.tasks.vision.facelandmarker.FaceLandmarker
import com.google.mediapipe.tasks.vision.facelandmarker.FaceLandmarkerResult
import com.google.mediapipe.tasks.vision.handlandmarker.HandLandmarker
import com.google.mediapipe.tasks.vision.handlandmarker.HandLandmarkerResult
import com.google.mediapipe.tasks.vision.imagesegmenter.ImageSegmenter
import com.google.mediapipe.tasks.vision.imagesegmenter.ImageSegmenterResult
import com.margelo.nitro.NitroModules
import com.margelo.nitro.camera.HybridFrameSpec
import com.margelo.nitro.camera.extensions.toBitmap
import com.margelo.nitro.camera.public.NativeFrame
import java.nio.ByteOrder
import kotlin.math.max
import kotlin.math.min
import kotlin.math.roundToInt

class HybridDrapeHandLandmarker : HybridDrapeHandLandmarkerSpec() {
  private var handLandmarker: HandLandmarker? = null

  init {
    DrapeVisionOnLoad.initializeNative()
  }

  override fun initialize(): Boolean {
    landmarkerFor(null)
    return true
  }

  override fun detectHands(
    frame: HybridFrameSpec,
    options: DrapeHandDetectionOptions,
  ): VisionHandDetectionResult {
    require(frame.isValid) { "Hand/Wrist Scan received an expired camera frame." }
    val nativeFrame = frame as? NativeFrame
      ?: throw IllegalArgumentException("Hand/Wrist Scan requires a native VisionCamera frame on Android.")

    val bitmap = nativeFrame.image.toBitmap(frame.orientation, frame.isMirrored)
    val mpImage = BitmapImageBuilder(bitmap).build()

    try {
      val timestampMs = frameTimestampMs(frame)
      val startNs = System.nanoTime()
      val result = landmarkerFor(options).detect(mpImage)
      val inferenceMs = elapsedMs(startNs)

      return VisionHandDetectionResult(
        hands = result.toVisionHands(),
        timestampMs = timestampMs,
        inferenceMs = inferenceMs,
        model = VisionSpecialistModel.HAND_LANDMARKER,
      )
    } finally {
      mpImage.close()
      bitmap.recycleIfNeeded()
    }
  }

  override fun clear() {
    handLandmarker?.close()
    handLandmarker = null
  }

  private fun landmarkerFor(options: DrapeHandDetectionOptions?): HandLandmarker {
    handLandmarker?.let { return it }
    val context = requireApplicationContext("Hand/Wrist Scan")
    val modelAssetPath = requireAsset(context, DrapeVisionNativeConstants.handModelAsset, "Hand/Wrist Scan")
    val handOptions = HandLandmarker.HandLandmarkerOptions.builder()
      .setBaseOptions(baseOptions(modelAssetPath))
      .setRunningMode(RunningMode.IMAGE)
      .setNumHands((options?.maxHands ?: 2.0).roundToInt().coerceIn(1, 2))
      .setMinHandDetectionConfidence(
        (options?.minHandDetectionConfidence?.toFloat() ?: DrapeVisionNativeConstants.defaultConfidence).coerceIn(0f, 1f)
      )
      .setMinHandPresenceConfidence(
        (options?.minHandPresenceConfidence?.toFloat() ?: DrapeVisionNativeConstants.defaultConfidence).coerceIn(0f, 1f)
      )
      .setMinTrackingConfidence(
        (options?.minTrackingConfidence?.toFloat() ?: DrapeVisionNativeConstants.defaultConfidence).coerceIn(0f, 1f)
      )
      .build()
    val created = HandLandmarker.createFromOptions(context, handOptions)
    handLandmarker = created
    return created
  }
}

class HybridDrapeFaceLandmarker : HybridDrapeFaceLandmarkerSpec() {
  private var faceLandmarker: FaceLandmarker? = null

  init {
    DrapeVisionOnLoad.initializeNative()
  }

  override fun initialize(): Boolean {
    landmarkerFor(null)
    return true
  }

  override fun detectFace(
    frame: HybridFrameSpec,
    options: DrapeFaceDetectionOptions,
  ): VisionFaceDetectionResult {
    require(frame.isValid) { "Headwear Scan received an expired camera frame." }
    val nativeFrame = frame as? NativeFrame
      ?: throw IllegalArgumentException("Headwear Scan requires a native VisionCamera frame on Android.")

    val bitmap = nativeFrame.image.toBitmap(frame.orientation, frame.isMirrored)
    val mpImage = BitmapImageBuilder(bitmap).build()

    try {
      val timestampMs = frameTimestampMs(frame)
      val startNs = System.nanoTime()
      val result = landmarkerFor(options).detect(mpImage)
      val inferenceMs = elapsedMs(startNs)
      val landmarks = result.faceLandmarks().firstOrNull().orEmpty()
      val blendshapes = result.toVisionBlendshapes()

      return VisionFaceDetectionResult(
        landmarks = landmarks.map { it.toVisionLandmark() }.toTypedArray(),
        blendshapes = blendshapes,
        faceCount = result.faceLandmarks().size.toDouble(),
        timestampMs = timestampMs,
        inferenceMs = inferenceMs,
        model = VisionSpecialistModel.FACE_LANDMARKER,
      )
    } finally {
      mpImage.close()
      bitmap.recycleIfNeeded()
    }
  }

  override fun clear() {
    faceLandmarker?.close()
    faceLandmarker = null
  }

  private fun landmarkerFor(options: DrapeFaceDetectionOptions?): FaceLandmarker {
    faceLandmarker?.let { return it }
    val context = requireApplicationContext("Headwear Scan")
    val modelAssetPath = requireAsset(context, DrapeVisionNativeConstants.faceModelAsset, "Headwear Scan")
    val faceOptions = FaceLandmarker.FaceLandmarkerOptions.builder()
      .setBaseOptions(baseOptions(modelAssetPath))
      .setRunningMode(RunningMode.IMAGE)
      .setNumFaces(1)
      .setMinFaceDetectionConfidence(
        (options?.minFaceDetectionConfidence?.toFloat() ?: DrapeVisionNativeConstants.defaultConfidence).coerceIn(0f, 1f)
      )
      .setMinFacePresenceConfidence(
        (options?.minFacePresenceConfidence?.toFloat() ?: DrapeVisionNativeConstants.defaultConfidence).coerceIn(0f, 1f)
      )
      .setMinTrackingConfidence(
        (options?.minTrackingConfidence?.toFloat() ?: DrapeVisionNativeConstants.defaultConfidence).coerceIn(0f, 1f)
      )
      .setOutputFaceBlendshapes(options?.outputFaceBlendshapes ?: false)
      .setOutputFacialTransformationMatrixes(false)
      .build()
    val created = FaceLandmarker.createFromOptions(context, faceOptions)
    faceLandmarker = created
    return created
  }
}

class HybridDrapeImageSegmenter : HybridDrapeImageSegmenterSpec() {
  private var imageSegmenter: ImageSegmenter? = null

  init {
    DrapeVisionOnLoad.initializeNative()
  }

  override fun initialize(): Boolean {
    segmenterFor(null)
    return true
  }

  override fun segment(
    frame: HybridFrameSpec,
    options: DrapeImageSegmentationOptions,
  ): VisionSegmentationResult {
    require(frame.isValid) { "Image Segmenter received an expired camera frame." }
    val nativeFrame = frame as? NativeFrame
      ?: throw IllegalArgumentException("Image Segmenter requires a native VisionCamera frame on Android.")

    val bitmap = nativeFrame.image.toBitmap(frame.orientation, frame.isMirrored)
    val mpImage = BitmapImageBuilder(bitmap).build()

    try {
      val timestampMs = frameTimestampMs(frame)
      val startNs = System.nanoTime()
      val result = segmenterFor(options).segment(mpImage)
      val inferenceMs = elapsedMs(startNs)

      return VisionSegmentationResult(
        mask = result.toVisionMaskSummary(options.confidenceThreshold ?: 0.5),
        timestampMs = timestampMs,
        inferenceMs = inferenceMs,
        model = VisionSpecialistModel.IMAGE_SEGMENTER,
      )
    } finally {
      mpImage.close()
      bitmap.recycleIfNeeded()
    }
  }

  override fun clear() {
    imageSegmenter?.close()
    imageSegmenter = null
  }

  private fun segmenterFor(options: DrapeImageSegmentationOptions?): ImageSegmenter {
    imageSegmenter?.let { return it }
    val context = requireApplicationContext("Image Segmenter")
    val modelAssetPath = requireAsset(context, DrapeVisionNativeConstants.imageSegmenterModelAsset, "Image Segmenter")
    val segmenterOptions = ImageSegmenter.ImageSegmenterOptions.builder()
      .setBaseOptions(baseOptions(modelAssetPath))
      .setRunningMode(RunningMode.IMAGE)
      .setDisplayNamesLocale("en")
      .setOutputConfidenceMasks(options?.outputConfidenceMasks ?: true)
      .setOutputCategoryMask(false)
      .build()
    val created = ImageSegmenter.createFromOptions(context, segmenterOptions)
    imageSegmenter = created
    return created
  }
}

private fun requireApplicationContext(module: String): Context {
  return NitroModules.applicationContext
    ?: throw IllegalStateException("$module cannot initialize before NitroModules has an application context.")
}

private fun requireAsset(context: Context, assetName: String, module: String): String {
  val candidatePaths = listOf(assetName, "models/$assetName")
  for (candidatePath in candidatePaths) {
    try {
      context.assets.open(candidatePath).use { }
      return candidatePath
    } catch (_: Throwable) {
      // Try the next candidate path.
    }
  }

  throw IllegalStateException(
    "$module model asset missing: $assetName. Rebuild and reinstall the app so the bundled Drape Vision model assets are copied into this build."
  )
}

private fun baseOptions(assetName: String): BaseOptions {
  return BaseOptions.builder()
    .setModelAssetPath(assetName)
    .setDelegate(Delegate.CPU)
    .build()
}

private fun frameTimestampMs(frame: HybridFrameSpec): Double {
  return frame.timestamp / 1_000_000.0
}

private fun elapsedMs(startNs: Long): Double {
  return (System.nanoTime() - startNs).toDouble() / 1_000_000.0
}

private fun HandLandmarkerResult.toVisionHands(): Array<VisionHandDetection> {
  return landmarks().mapIndexed { index, handLandmarks ->
    val handWorldLandmarks = worldLandmarks().getOrNull(index)
    val handednessCategory = handedness().getOrNull(index)?.maxByOrNull { it.score() }
    VisionHandDetection(
      landmarks = handLandmarks.map { it.toVisionLandmark() }.toTypedArray(),
      worldLandmarks = handWorldLandmarks?.map { it.toVisionLandmark() }?.toTypedArray(),
      handedness = handednessCategory?.toVisionHandedness(),
      confidence = handednessCategory?.score()?.toDouble(),
    )
  }.toTypedArray()
}

private fun FaceLandmarkerResult.toVisionBlendshapes(): Array<VisionFaceBlendshape>? {
  val firstFaceBlendshapes = faceBlendshapes().orElse(null)?.firstOrNull() ?: return null
  return firstFaceBlendshapes.map { category ->
    VisionFaceBlendshape(
      categoryName = category.categoryName() ?: category.displayName() ?: "unknown",
      score = category.score().toDouble(),
    )
  }.toTypedArray()
}

private fun ImageSegmenterResult.toVisionMaskSummary(confidenceThreshold: Double): VisionSegmentationMaskSummary? {
  val confidenceMask = confidenceMasks().orElse(null)?.firstOrNull()
  if (confidenceMask != null) {
    return summarizeMask(confidenceMask, confidenceThreshold.coerceIn(0.0, 1.0).toFloat())
  }

  val categoryMask = categoryMask().orElse(null) ?: return null
  return summarizeMask(categoryMask, 1f)
}

private fun summarizeMask(maskImage: MPImage, confidenceThreshold: Float): VisionSegmentationMaskSummary? {
  val bitmap = try {
    BitmapExtractor.extract(maskImage)
  } catch (_: Throwable) {
    null
  }
  if (bitmap != null) {
    return summarizeBitmapMask(bitmap, confidenceThreshold)
  }

  val floatBuffer = try {
    ByteBufferExtractor.extract(maskImage, MPImage.IMAGE_FORMAT_VEC32F1).order(ByteOrder.nativeOrder()).asFloatBuffer()
  } catch (_: Throwable) {
    return null
  }
  val width = maskImage.width
  val height = maskImage.height
  if (width <= 0 || height <= 0 || floatBuffer.capacity() < width * height) return null

  var foregroundCount = 0
  var minX = width
  var maxX = -1
  var minY = height
  var maxY = -1
  for (y in 0 until height) {
    for (x in 0 until width) {
      val confidence = floatBuffer.get(y * width + x)
      if (confidence >= confidenceThreshold) {
        foregroundCount += 1
        minX = min(minX, x)
        maxX = max(maxX, x)
        minY = min(minY, y)
        maxY = max(maxY, y)
      }
    }
  }
  return maskSummary(width, height, foregroundCount, minX, maxX, minY, maxY)
}

private fun summarizeBitmapMask(bitmap: Bitmap, confidenceThreshold: Float): VisionSegmentationMaskSummary? {
  val width = bitmap.width
  val height = bitmap.height
  if (width <= 0 || height <= 0) return null

  val threshold = (confidenceThreshold.coerceIn(0f, 1f) * 255f).roundToInt()
  var foregroundCount = 0
  var minX = width
  var maxX = -1
  var minY = height
  var maxY = -1
  for (y in 0 until height) {
    for (x in 0 until width) {
      val alpha = bitmap.getPixel(x, y) ushr 24
      if (alpha >= threshold) {
        foregroundCount += 1
        minX = min(minX, x)
        maxX = max(maxX, x)
        minY = min(minY, y)
        maxY = max(maxY, y)
      }
    }
  }
  return maskSummary(width, height, foregroundCount, minX, maxX, minY, maxY)
}

private fun maskSummary(
  width: Int,
  height: Int,
  foregroundCount: Int,
  minX: Int,
  maxX: Int,
  minY: Int,
  maxY: Int,
): VisionSegmentationMaskSummary {
  val pixelCount = (width * height).coerceAtLeast(1)
  val boundingBox = if (foregroundCount > 0 && maxX >= minX && maxY >= minY) {
    VisionBoundingBox(
      x = minX.toDouble(),
      y = minY.toDouble(),
      width = (maxX - minX + 1).toDouble(),
      height = (maxY - minY + 1).toDouble(),
    )
  } else {
    null
  }
  return VisionSegmentationMaskSummary(
    width = width.toDouble(),
    height = height.toDouble(),
    foregroundRatio = foregroundCount.toDouble() / pixelCount.toDouble(),
    boundingBox = boundingBox,
  )
}

private fun NormalizedLandmark.toVisionLandmark(): VisionLandmark {
  return VisionLandmark(
    x = x().toDouble(),
    y = y().toDouble(),
    z = z().toDouble(),
    visibility = visibility().orElse(null)?.toDouble(),
    presence = presence().orElse(null)?.toDouble(),
  )
}

private fun Landmark.toVisionLandmark(): VisionLandmark {
  return VisionLandmark(
    x = x().toDouble(),
    y = y().toDouble(),
    z = z().toDouble(),
    visibility = visibility().orElse(null)?.toDouble(),
    presence = presence().orElse(null)?.toDouble(),
  )
}

private fun Category.toVisionHandedness(): VisionHandedness {
  return when ((categoryName() ?: displayName() ?: "").uppercase()) {
    "LEFT" -> VisionHandedness.LEFT
    "RIGHT" -> VisionHandedness.RIGHT
    else -> VisionHandedness.UNKNOWN
  }
}

private fun Bitmap.recycleIfNeeded() {
  if (!isRecycled) recycle()
}

File: packages/drape-vision/ios/HybridDrapeSpecialistLandmarkers.swift
import AVFoundation
import Foundation
import NitroModules
import UIKit
import VisionCamera

#if canImport(MediaPipeTasksVision)
import MediaPipeTasksVision
#endif

final class HybridDrapeHandLandmarker: HybridDrapeHandLandmarkerSpec {
  #if canImport(MediaPipeTasksVision)
  private var handLandmarker: HandLandmarker?
  #endif

  func initialize() throws -> Bool {
    #if canImport(MediaPipeTasksVision)
    _ = try landmarker(options: nil)
    return true
    #else
    throw RuntimeError.error(withMessage: "Hand/Wrist Scan requires the MediaPipeTasksVision CocoaPod.")
    #endif
  }

  func detectHands(frame: any HybridFrameSpec, options: DrapeHandDetectionOptions) throws -> VisionHandDetectionResult {
    #if canImport(MediaPipeTasksVision)
    let sampleBuffer = try sampleBuffer(from: frame, module: "Hand/Wrist Scan")
    let timestampMs = timestampMs(from: frame)
    let image = try MPImage(sampleBuffer: sampleBuffer, orientation: imageOrientation(for: frame))
    let startedAt = Date()
    let result = try landmarker(options: options).detect(image: image)
    let inferenceMs = Date().timeIntervalSince(startedAt) * 1000

    return VisionHandDetectionResult(
      hands: result.toVisionHands(),
      timestampMs: timestampMs,
      inferenceMs: inferenceMs,
      model: .handLandmarker
    )
    #else
    throw RuntimeError.error(withMessage: "Hand/Wrist Scan requires the MediaPipeTasksVision CocoaPod.")
    #endif
  }

  func clear() throws {
    #if canImport(MediaPipeTasksVision)
    handLandmarker = nil
    #endif
  }

  #if canImport(MediaPipeTasksVision)
  private func landmarker(options detectionOptions: DrapeHandDetectionOptions?) throws -> HandLandmarker {
    if let handLandmarker {
      return handLandmarker
    }

    let options = HandLandmarkerOptions()
    options.baseOptions.modelAssetPath = try modelPath(
      resourceName: DrapeVisionNativeConstants.handModelResourceName,
      module: "Hand/Wrist Scan"
    )
    options.runningMode = .image
    options.numHands = Int(detectionOptions?.maxHands ?? 2).clamped(to: 1...2)
    options.minHandDetectionConfidence = detectionOptions?.minHandDetectionConfidence.map(Float.init)
      ?? DrapeVisionNativeConstants.defaultConfidenceThreshold
    options.minHandPresenceConfidence = detectionOptions?.minHandPresenceConfidence.map(Float.init)
      ?? DrapeVisionNativeConstants.defaultConfidenceThreshold
    options.minTrackingConfidence = detectionOptions?.minTrackingConfidence.map(Float.init)
      ?? DrapeVisionNativeConstants.defaultConfidenceThreshold

    let created = try HandLandmarker(options: options)
    handLandmarker = created
    return created
  }
  #endif
}

final class HybridDrapeFaceLandmarker: HybridDrapeFaceLandmarkerSpec {
  #if canImport(MediaPipeTasksVision)
  private var faceLandmarker: FaceLandmarker?
  #endif

  func initialize() throws -> Bool {
    #if canImport(MediaPipeTasksVision)
    _ = try landmarker(options: nil)
    return true
    #else
    throw RuntimeError.error(withMessage: "Headwear Scan requires the MediaPipeTasksVision CocoaPod.")
    #endif
  }

  func detectFace(frame: any HybridFrameSpec, options: DrapeFaceDetectionOptions) throws -> VisionFaceDetectionResult {
    #if canImport(MediaPipeTasksVision)
    let sampleBuffer = try sampleBuffer(from: frame, module: "Headwear Scan")
    let timestampMs = timestampMs(from: frame)
    let image = try MPImage(sampleBuffer: sampleBuffer, orientation: imageOrientation(for: frame))
    let startedAt = Date()
    let result = try landmarker(options: options).detect(image: image)
    let inferenceMs = Date().timeIntervalSince(startedAt) * 1000
    let landmarks = result.faceLandmarks.first ?? []

    return VisionFaceDetectionResult(
      landmarks: landmarks.map(toVisionLandmark),
      blendshapes: result.toVisionBlendshapes(),
      faceCount: Double(result.faceLandmarks.count),
      timestampMs: timestampMs,
      inferenceMs: inferenceMs,
      model: .faceLandmarker
    )
    #else
    throw RuntimeError.error(withMessage: "Headwear Scan requires the MediaPipeTasksVision CocoaPod.")
    #endif
  }

  func clear() throws {
    #if canImport(MediaPipeTasksVision)
    faceLandmarker = nil
    #endif
  }

  #if canImport(MediaPipeTasksVision)
  private func landmarker(options detectionOptions: DrapeFaceDetectionOptions?) throws -> FaceLandmarker {
    if let faceLandmarker {
      return faceLandmarker
    }

    let options = FaceLandmarkerOptions()
    options.baseOptions.modelAssetPath = try modelPath(
      resourceName: DrapeVisionNativeConstants.faceModelResourceName,
      module: "Headwear Scan"
    )
    options.runningMode = .image
    options.numFaces = 1
    options.minFaceDetectionConfidence = detectionOptions?.minFaceDetectionConfidence.map(Float.init)
      ?? DrapeVisionNativeConstants.defaultConfidenceThreshold
    options.minFacePresenceConfidence = detectionOptions?.minFacePresenceConfidence.map(Float.init)
      ?? DrapeVisionNativeConstants.defaultConfidenceThreshold
    options.minTrackingConfidence = detectionOptions?.minTrackingConfidence.map(Float.init)
      ?? DrapeVisionNativeConstants.defaultConfidenceThreshold
    options.outputFaceBlendshapes = detectionOptions?.outputFaceBlendshapes ?? false
    options.outputFacialTransformationMatrixes = false

    let created = try FaceLandmarker(options: options)
    faceLandmarker = created
    return created
  }
  #endif
}

final class HybridDrapeImageSegmenter: HybridDrapeImageSegmenterSpec {
  #if canImport(MediaPipeTasksVision)
  private var imageSegmenter: ImageSegmenter?
  #endif

  func initialize() throws -> Bool {
    #if canImport(MediaPipeTasksVision)
    _ = try segmenter(options: nil)
    return true
    #else
    throw RuntimeError.error(withMessage: "Image Segmenter requires the MediaPipeTasksVision CocoaPod.")
    #endif
  }

  func segment(frame: any HybridFrameSpec, options: DrapeImageSegmentationOptions) throws -> VisionSegmentationResult {
    #if canImport(MediaPipeTasksVision)
    let sampleBuffer = try sampleBuffer(from: frame, module: "Image Segmenter")
    let timestampMs = timestampMs(from: frame)
    let image = try MPImage(sampleBuffer: sampleBuffer, orientation: imageOrientation(for: frame))
    let startedAt = Date()
    let result = try segmenter(options: options).segment(image: image)
    let inferenceMs = Date().timeIntervalSince(startedAt) * 1000

    return VisionSegmentationResult(
      mask: result.toVisionMaskSummary(confidenceThreshold: options.confidenceThreshold ?? 0.5),
      timestampMs: timestampMs,
      inferenceMs: inferenceMs,
      model: .imageSegmenter
    )
    #else
    throw RuntimeError.error(withMessage: "Image Segmenter requires the MediaPipeTasksVision CocoaPod.")
    #endif
  }

  func clear() throws {
    #if canImport(MediaPipeTasksVision)
    imageSegmenter = nil
    #endif
  }

  #if canImport(MediaPipeTasksVision)
  private func segmenter(options detectionOptions: DrapeImageSegmentationOptions?) throws -> ImageSegmenter {
    if let imageSegmenter {
      return imageSegmenter
    }

    let options = ImageSegmenterOptions()
    options.baseOptions.modelAssetPath = try modelPath(
      resourceName: DrapeVisionNativeConstants.imageSegmenterModelResourceName,
      module: "Image Segmenter"
    )
    options.runningMode = .image
    options.displayNamesLocale = "en"
    options.shouldOutputConfidenceMasks = detectionOptions?.outputConfidenceMasks ?? true
    options.shouldOutputCategoryMask = false

    let created = try ImageSegmenter(options: options)
    imageSegmenter = created
    return created
  }
  #endif
}

#if canImport(MediaPipeTasksVision)
private func sampleBuffer(from frame: any HybridFrameSpec, module: String) throws -> CMSampleBuffer {
  guard let nativeFrame = frame as? any NativeFrame else {
    throw RuntimeError.error(withMessage: "\(module) expected a VisionCamera native Frame.")
  }
  guard let sampleBuffer = nativeFrame.sampleBuffer else {
    throw RuntimeError.error(withMessage: "\(module) cannot read a disposed camera frame.")
  }
  return sampleBuffer
}

private func timestampMs(from frame: any HybridFrameSpec) -> Double {
  frame.timestamp * 1000
}

private func imageOrientation(for frame: any HybridFrameSpec) -> UIImage.Orientation {
  switch frame.orientation {
  case .up:
    return .up
  case .right:
    return .right
  case .down:
    return .down
  case .left:
    return .left
  }
}

private func modelPath(resourceName: String, module: String) throws -> String {
  let candidateBundles = [Bundle.main, Bundle(for: HybridDrapePoseLandmarker.self)] + Bundle.allBundles + Bundle.allFrameworks
  for bundle in candidateBundles {
    if let path = bundle.path(forResource: resourceName, ofType: DrapeVisionNativeConstants.modelResourceExtension) {
      return path
    }
    if let path = bundle.path(
      forResource: resourceName,
      ofType: DrapeVisionNativeConstants.modelResourceExtension,
      inDirectory: "models"
    ) {
      return path
    }
  }

  throw RuntimeError.error(
    withMessage: "\(module) model asset missing: \(resourceName).\(DrapeVisionNativeConstants.modelResourceExtension). Rebuild and reinstall the app so the bundled Drape Vision model assets are copied into this build."
  )
}

private func toVisionLandmark(_ landmark: NormalizedLandmark) -> VisionLandmark {
  VisionLandmark(
    x: Double(landmark.x),
    y: Double(landmark.y),
    z: Double(landmark.z),
    visibility: landmark.visibility?.doubleValue,
    presence: landmark.presence?.doubleValue
  )
}

private func toVisionLandmark(_ landmark: Landmark) -> VisionLandmark {
  VisionLandmark(
    x: Double(landmark.x),
    y: Double(landmark.y),
    z: Double(landmark.z),
    visibility: landmark.visibility?.doubleValue,
    presence: landmark.presence?.doubleValue
  )
}

private func summarizeMask(_ mask: Mask, confidenceThreshold: Float) -> VisionSegmentationMaskSummary? {
  guard mask.width > 0, mask.height > 0 else { return nil }
  let data = mask.float32Data
  let width = mask.width
  let height = mask.height
  var foregroundCount = 0
  var minX = width
  var maxX = -1
  var minY = height
  var maxY = -1

  for y in 0..<height {
    for x in 0..<width {
      if data[y * width + x] >= confidenceThreshold {
        foregroundCount += 1
        minX = min(minX, x)
        maxX = max(maxX, x)
        minY = min(minY, y)
        maxY = max(maxY, y)
      }
    }
  }

  let boundingBox: VisionBoundingBox?
  if foregroundCount > 0, maxX >= minX, maxY >= minY {
    boundingBox = VisionBoundingBox(
      x: Double(minX),
      y: Double(minY),
      width: Double(maxX - minX + 1),
      height: Double(maxY - minY + 1)
    )
  } else {
    boundingBox = nil
  }

  return VisionSegmentationMaskSummary(
    width: Double(width),
    height: Double(height),
    foregroundRatio: Double(foregroundCount) / Double(max(width * height, 1)),
    boundingBox: boundingBox
  )
}

private extension HandLandmarkerResult {
  func toVisionHands() -> [VisionHandDetection] {
    landmarks.enumerated().map { index, handLandmarks in
      let handednessCategory = handedness[safe: index]?.max(by: { $0.score < $1.score })
      let worldLandmarksForHand = worldLandmarks[safe: index]
      return VisionHandDetection(
        landmarks: handLandmarks.map(toVisionLandmark),
        worldLandmarks: worldLandmarksForHand?.map(toVisionLandmark),
        handedness: handednessCategory?.visionHandedness,
        confidence: handednessCategory.map { Double($0.score) }
      )
    }
  }
}

private extension FaceLandmarkerResult {
  func toVisionBlendshapes() -> [VisionFaceBlendshape]? {
    guard let classifications = faceBlendshapes.first else {
      return nil
    }
    return classifications.categories.map {
      VisionFaceBlendshape(
        categoryName: $0.categoryName ?? $0.displayName ?? "unknown",
        score: Double($0.score)
      )
    }
  }
}

private extension ImageSegmenterResult {
  func toVisionMaskSummary(confidenceThreshold: Double) -> VisionSegmentationMaskSummary? {
    if let confidenceMask = confidenceMasks?.first {
      return summarizeMask(confidenceMask, confidenceThreshold: Float(confidenceThreshold).clamped(to: 0...1))
    }
    if let categoryMask {
      return summarizeMask(categoryMask, confidenceThreshold: 1)
    }
    return nil
  }
}

private extension ResultCategory {
  var visionHandedness: VisionHandedness {
    switch (categoryName ?? displayName ?? "").uppercased() {
    case "LEFT":
      return .left
    case "RIGHT":
      return .right
    default:
      return .unknown
    }
  }
}

private extension Collection {
  subscript(safe index: Index) -> Element? {
    indices.contains(index) ? self[index] : nil
  }
}

private extension Comparable {
  func clamped(to range: ClosedRange<Self>) -> Self {
    min(max(self, range.lowerBound), range.upperBound)
  }
}
#endif
